src/stream/src/from_proto/chain.rs

```diff
@@ -43,26 +50,98 @@ impl ExecutorBuilder for ChainExecutorBuilder {
                 schema,
                 params.pk_indices,
-            );
-            Ok(executor.boxed())
-        } else {
-            let executor = RearrangedChainExecutor::new(
+            )
+            .boxed(),
+            ChainType::Rearrange => RearrangedChainExecutor::new(
                 snapshot,
                 mview,
                 upstream_indices,
                 progress,
                 schema,
                 params.pk_indices,
-            );
-            Ok(executor.boxed())
-        }
+            )
+            .boxed(),
+            ChainType::Backfill => {
+                let table_desc: &StorageTableDesc = node.get_table_desc()?;
+                let table_id = TableId {
+                    table_id: table_desc.table_id,
+                };
+
+                let order_types = table_desc
+                    .pk
+                    .iter()
+                    .map(|desc| {
+                        OrderType::from_prost(&ProstOrderType::from_i32(desc.order_type).unwrap())
+                    })
+                    .collect_vec();
+
+                let column_descs = table_desc
+                    .columns
+                    .iter()
+                    .map(ColumnDesc::from)
+                    .collect_vec();
+                let column_ids = column_descs.iter().map(|x| x.column_id).collect_vec();
+
+                // Use indices based on full table instead of streaming executor output.
+                let pk_indices = table_desc.pk.iter().map(|k| k.index as usize).collect_vec();
+
+                let dist_key_indices = table_desc
+                    .dist_key_indices
+                    .iter()
+                    .map(|&k| k as usize)
+                    .collect_vec();
+                let distribution = match params.vnode_bitmap {
+                    Some(vnodes) => Distribution {
+                        dist_key_indices,
+                        vnodes: vnodes.into(),
+                    },
+                    None => Distribution::fallback(),
+                };
+
+                let table_option = TableOption {
+                    retention_seconds: if table_desc.retention_seconds > 0 {
+                        Some(table_desc.retention_seconds)
+                    } else {
+                        None
+                    },
+                };
+                let value_indices = table_desc
+                    .get_value_indices()
+                    .iter()
+                    .map(|&k| k as usize)
+                    .collect_vec();
+                let table = StorageTable::new_partial(
+                    state_store,
+                    table_id,
+                    column_descs,
+                    column_ids,
+                    order_types,
+                    pk_indices,
+                    distribution,
+                    table_option,
+                    value_indices,
+                );
+
+                BackfillExecutor::new(
```

Comment:
Previously we expected the `BatchQueryExecutor` to be more general to support some simple pruning, but it seems to be unused by the `Backfill`. I'm not sure what the future plan is, should we keep that, or totally remove that and try something like what we did in `BatchLookupJoin`?

Reply:
I think currently BatchQueryExecutor is only used by the `StreamIndexScan`. I haven't figured out why it use `ChainExecutor` instead of `RearrangedChainExecutor` before. Once we can replace all of them by `BackfillExecutor` I think we don't need the `BatchQueryExecutor `. However, once we remove the `BatchQuery` from the `ChainNode`, it seems like `ChainNode` is not a proper name.